pandas/tests/test_strings.py

```diff
@@ -34,12 +150,67 @@ def test_api(self):
         assert Series.str is strings.StringMethods
         assert isinstance(Series(['']).str, strings.StringMethods)
 
-        # GH 9184
-        invalid = Series([1])
+    @pytest.mark.parametrize('box', [Series, Index])
+    def test_api_per_dtype(self, box, all_skipna_inferred_dtypes):
+        # one instance of parametrized fixture
+        inferred_type, values = all_skipna_inferred_dtypes
+
+        t = box(values, dtype=object)  # object dtype to avoid casting
+
+        types_passing_constructor = ['string', 'unicode', 'empty',
+                                     'bytes', 'mixed', 'mixed-integer']
+        if inferred_type in types_passing_constructor:
+            # GH 6106, GH 9322
+            assert isinstance(t.str, strings.StringMethods)
+        else:
+            # GH 9184, GH 23011, GH 23163
+            with tm.assert_raises_regex(AttributeError,
+                                        "Can only use .str accessor with "
+                                        "string values!"):
+                t.str
+            assert not hasattr(t, 'str')
+
+    def test_api_mi_raises(self):
```

Comment:
As far as I can tell, the (existing) error for trying to call `MultiIndex.str` was not tested so far.